Caret painting should listen for change notifications

mwyrzykowski opened the pull request:
#### 67d841b77448d21fea79f6cff02908859b307eca
<pre>
Caret painting should listen for change notifications
<a href="https://bugs.webkit.org/show_bug.cgi?id=252879">https://bugs.webkit.org/show_bug.cgi?id=252879</a>
&lt;radar://105862690&gt;

Reviewed by Aditya Keerthi.

Addition refactoring that listens to notifications from the UI process.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::repaintCaretForLocalRect):
(WebCore::CaretBase::paintCaret const):
(WebCore::FrameSelection::caretAnimatorInvalidated):
Repaint the caret when it changes.

(WebCore::fillCaretRect): Deleted.
(WebCore::repaintCaretRectForLocalRect): Deleted.
Move some static functions into CaretAnimator.

* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/platform/CaretAnimator.cpp:
(WebCore::CaretAnimator::paint const):
(WebCore::CaretAnimator::repaintCaretRectForLocalRect const):
Move some static functions into CaretAnimator.

* Source/WebCore/platform/CaretAnimator.h:
(WebCore::CaretAnimationClient::caretAnimationDidUpdate):
(WebCore::CaretAnimationClient::caretAnimatorInvalidated): Deleted.
Renamed.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setCaretAppearance):
Update the caret.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::WebViewImpl):
Listen for notifications.

* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::insertDictatedTextAsync):
This assertion is regularly triggered and igorning it results
in duplicated text insertion during dictation, so early return
when it is encountered which results in what appears to be correct
behavior.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setCaretAppearance):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
Allow UI process to signal caret appearance updates.

Canonical link: <a href="https://commits.webkit.org/261089@main">https://commits.webkit.org/261089@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b2c5859fff0535e1bc8fd8c05d55a058ef329e4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110542 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19624 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/43122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/1904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114491 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21053 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/10764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102774 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116284 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/43122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/113959 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/43122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12279 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/43122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12849 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/10764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18213 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/43122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7677 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14722 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
```diff
@@ -506,6 +506,7 @@ GenerateSyntheticEditingCommand(enum:uint8_t WebKit::SyntheticEditingCommandType
 #if PLATFORM(MAC)
     AttributedSubstringForCharacterRangeAsync(struct WebKit::EditingRange range) -> (struct WebCore::AttributedString string, struct WebKit::EditingRange range)
     RequestAcceptsFirstMouse(int eventNumber, WebKit::WebMouseEvent event) AllowedWhenWaitingForSyncReplyDuringUnboundedIPC
+    SetCaretAppearance(int appearance)
```
Would be nicer if this took a `CaretAnimatorType` parameter. Should be doable by adding something to `WebCoreArgumentCoders.serialization.in`.

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -3152,6 +3156,14 @@ static String commandNameForSelector(SEL selector)
+#ifndef WEB_VIEW_IMPL_ADDITIONS_SUSCRIBE_TO_TEXT_INPUT_NOTIFICATIONS_DEFINED
+static RetainPtr<NSObject> subscribeToTextInputNotifications(WebViewImpl*);
+{
+    return adoptNS([[NSObject alloc] init]);
```
Maybe better to just return `nil` if possible?

pxlcoder commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -29,23 +29,25 @@
-enum class CaretAnimatorType {
-    Default,
-    Alternate
-};
```
Is there a reason we're moving the enum out of here?

mwyrzykowski replied:
It is to avoid including CaretAnimator.h in WebPage.cpp and WebPageProxy.cpp

But maybe I can use a forward declaration of the enum class instead

mwyrzykowski committed:
Caret painting should listen for change notifications
https://bugs.webkit.org/show_bug.cgi?id=252879
<radar://105862690>

Reviewed by Aditya Keerthi.

Addition refactoring that listens to notifications from the UI process.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::repaintCaretForLocalRect):
(WebCore::CaretBase::paintCaret const):
(WebCore::FrameSelection::caretAnimatorInvalidated):
Repaint the caret when it changes.

(WebCore::fillCaretRect): Deleted.
(WebCore::repaintCaretRectForLocalRect): Deleted.
Move some static functions into CaretAnimator.

* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/platform/CaretAnimator.cpp:
(WebCore::CaretAnimator::paint const):
(WebCore::CaretAnimator::repaintCaretRectForLocalRect const):
Move some static functions into CaretAnimator.

* Source/WebCore/platform/CaretAnimator.h:
(WebCore::CaretAnimationClient::caretAnimationDidUpdate):
(WebCore::CaretAnimationClient::caretAnimatorInvalidated): Deleted.
Renamed.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setCaretAppearance):
Update the caret.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::WebViewImpl):
Listen for notifications.

* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::insertDictatedTextAsync):
This assertion is regularly triggered and igorning it results
in duplicated text insertion during dictation, so early return
when it is encountered which results in what appears to be correct
behavior.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setCaretAppearance):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
Allow UI process to signal caret appearance updates.

Canonical link: https://commits.webkit.org/261089@main
